src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,193 @@
+/* eslint-disable jsx-a11y/label-has-associated-control */
+import React, { useEffect, useRef, useState } from 'react';
+import cn from 'classnames';
+import { Todo } from '../../types/todo/Todo';
+
+const TODO_DIV_CLASSNAME = 'todo';
+const TODO_REMOVE_BUTTON_CLASSNAME = 'todo__remove';
+const TODO_STATUS_CHECKBOX_CLASSNAME = 'todo__status';
+
+type Props = {
+  todo: Todo;
+  isTodoLoading: (todoId: Todo['id']) => boolean;
+  isTempTodo?: boolean;
+  onTodoDelete?: (
+    todoId: Todo['id'],
+    isDeleteAfterUpdate?: boolean,
+  ) => Promise<void>;
+  onTodoUpdate?: (
+    todoId: Todo['id'],
+    todoData: Partial<Todo>,
+    isTitleChange?: boolean,
+  ) => Promise<void>;
+  editingTodoId?: Todo['id'] | null;
+  setEditingTodoId?: (todoId: Todo['id'] | null) => void;
+  isEditTodoFormFocused?: boolean;
+  setIsEditTodoFormFocused?: (isFocused: boolean) => void;
+};
+
+export const TodoItem: React.FC<Props> = ({
+  todo,
+  isTodoLoading,
+  isTempTodo = false,
+  onTodoDelete,
+  onTodoUpdate,
+  editingTodoId,
+  setEditingTodoId,
+  isEditTodoFormFocused,
+  setIsEditTodoFormFocused,
+}) => {
+  const [editingTitle, setEditingTitle] = useState(todo.title);
+  const editingTitleInputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (editingTodoId === todo.id) {
+      setEditingTitle(prevTitleValue =>
+        prevTitleValue === todo.title ? todo.title : prevTitleValue,
+      );
+      editingTitleInputRef.current?.focus();
+    }
+  }, [editingTodoId, todo.id, todo.title]);
+
+  useEffect(() => {
+    if (isEditTodoFormFocused && setIsEditTodoFormFocused) {
+      editingTitleInputRef.current?.focus();
+      setIsEditTodoFormFocused(false);
+    }
+  }, [isEditTodoFormFocused, setIsEditTodoFormFocused]);
+
+  const handleChangeTodoStatus = () => {
+    if (onTodoUpdate) {
+      onTodoUpdate(todo.id, { completed: !todo.completed });
+    }
+  };
+
+  const handleDeleteTodo = () => {
+    if (onTodoDelete) {
+      onTodoDelete(todo.id);
+    }
+  };
+
+  const handleDoubleClickOnTodo = (
+    event: React.MouseEvent<HTMLDivElement, MouseEvent>,
+  ) => {
+    const target = event.target as HTMLElement;
+
+    const isClickOnDeleteButton = target.closest(
+      `.${TODO_REMOVE_BUTTON_CLASSNAME}`,
+    );
+    const isClickOnCheckbox = target.closest(
+      `.${TODO_STATUS_CHECKBOX_CLASSNAME}`,
+    );
+
+    const isClickOnTodoDiv = (
+      event.currentTarget as HTMLElement
+    ).classList.contains(TODO_DIV_CLASSNAME);
```

Comment:
why? what is this? why we need this?
it's hard to understand this please can we get rid of this?
i believe we should KISS